Signal/src/network/GiphyAPI.swift

```diff
@@ -0,0 +1,437 @@
+    }
+}
+
+// Represents a single Giphy image.
+@objc class GiphyImageInfo: NSObject {
+    let giphyId: String
+    let renditions: [GiphyRendition]
+    // We special-case the "original" rendition because it is the 
+    // source of truth for the aspect ratio of the image.
+    let originalRendition: GiphyRendition
+
+    init(giphyId: String,
+         renditions: [GiphyRendition],
+         originalRendition: GiphyRendition) {
+        self.giphyId = giphyId
+        self.renditions = renditions
+        self.originalRendition = originalRendition
+    }
+
+    // TODO: We may need to tweak these constants.
+    let kMaxDimension = UInt(618)
+    let kMinDimension = UInt(101)
+    let kMaxFileSize = UInt(3 * 1024 * 1024)
+
+    private enum PickingStrategy {
+        case smallerIsBetter, largerIsBetter
+    }
+
+    public func log() {
+        Logger.verbose("giphyId: \(giphyId), \(renditions.count)")
+        for rendition in renditions {
+            rendition.log()
+        }
+    }
+
+    public func pickStillRendition() -> GiphyRendition? {
+        // Stills are just temporary placeholders, so use the smallest still possible.
+        return pickRendition(isStill:true, pickingStrategy:.smallerIsBetter, maxFileSize:kMaxFileSize)
+    }
+
+    public func pickGifRendition() -> GiphyRendition? {
+        // Try to pick a small file...
+        if let rendition = pickRendition(isStill:false, pickingStrategy:.largerIsBetter, maxFileSize:kMaxFileSize) {
+            return rendition
+        }
+        // ...but gradually relax the file restriction...
+        if let rendition = pickRendition(isStill:false, pickingStrategy:.smallerIsBetter, maxFileSize:kMaxFileSize * 2) {
+            return rendition
+        }
+        // ...and relax even more until we find an animated rendition.
+        return pickRendition(isStill:false, pickingStrategy:.smallerIsBetter, maxFileSize:kMaxFileSize * 3)
+    }
+
+    // Picking a rendition must be done very carefully.
+    //
+    // * We want to avoid incomplete renditions.
+    // * We want to pick a rendition of "just good enough" quality.
+    private func pickRendition(isStill: Bool, pickingStrategy: PickingStrategy, maxFileSize: UInt) -> GiphyRendition? {
+        var bestRendition: GiphyRendition?
+
+        for rendition in renditions {
+            if isStill {
+                // Accept GIF or JPEG stills.  In practice we'll
+                // usually select a JPEG since they'll be smaller.
+                guard [.gif, .jpg].contains(rendition.format) else {
+                    continue
+                }
+                // Only consider still renditions.
+                guard rendition.name.hasSuffix("_still") else {
+                        continue
+                }
+                // Accept renditions without a valid file size.
+                //
+                // Don't worry about max content size; still images are tiny in comparison
```

Comment:
> But we compare rendition.fileSize below. So it seems like we don't accept renditions without a valid file size. Or is there some other definition of valid?

I've elaborated the comment, but basically our parsing defaults fileSize to zero if not specified by the Giphy API, so this test will pass.

> By "content size" you mean "resolution", right?

This might be pedantic, but I try to avoid using "resolution" to refer to the width & height of an image because that term is overloaded in imaging.  For example, iOS uses "resolution" to refer to the ratio of screen points to pixel.